### DISCUSSION TITLE
How to target specific framework/core versions when building with CSharpCompilation

### DISCUSSION THREAD

**Discussion by leet101** [dotnet/roslyn#57028] [Q&A] [answered]
Hi, I am trying to create an application that can use CSharpCompilation to build DLLs in different versions of framework/core.

The application where I am creating the dll is .net framework 4.7.2

Whatever I try, the compiled dll ends up in .net framework 4.0, with references to
C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework.NETFramework\v4.0\mscorlib.dll

I have tried adding various different references to the project, mscorlib and all other references in the following folder
C:\Program Files (x86)\Reference Assemblies\Microsoft\Framework.NETFramework\v4.7.2\
But still it ends up as .net framework 4.0

Please can you have a look at my example code and see if you can help, any advice would be much appreciated.

       private void compileDll()
       {
           var syntaxTree = SyntaxFactory.ParseSyntaxTree(SourceText.From(programText));
           var assemblyPath = "mydll.dll";
      
           string refLoc = typeof(object).GetTypeInfo().Assembly.Location;
      
           var compilation = CSharpCompilation.Create(assemblyPath)
               .WithOptions(new CSharpCompilationOptions(OutputKind.DynamicallyLinkedLibrary))
               .AddReferences(MetadataReference.CreateFromFile(refLoc))
               .AddSyntaxTrees(syntaxTree);
      
           var result = compilation.Emit(assemblyPath);
      
           MessageBox.Show("Result = " + result.Success);
       }

[Program.cs.txt](https://github.com/dotnet/roslyn/files/7308651/Program.cs.txt)

**Comment by huoyaoyuan** [comment]
> But still it ends up as .net framework 4.0

How did you confirm that?

Note that ALL of .NET 4.X assemblies are versioned as 4.0.0.0, which make them compatible.

**Comment by leet101** [comment]
Hi, I am using JetBrains to examine the compiled dll.
![image](https://user-images.githubusercontent.com/92138113/137861831-47225192-6e65-4c70-babd-2673c6c558ec.png)

If I compile a class lib as .net framework 4.7.2, this shows the following in JetBrains 
![image](https://user-images.githubusercontent.com/92138113/137862436-6eefec45-9c0a-476e-a10e-e70144d1a9ca.png)


I was expecting them to both show the same version.
I will mention I am new to using CSharpCompilation, please let me know if I am missing anything.

**Reply by svick** [reply]
It seems JetBrains displays whatever is in the `TargetFramework` attribute. E.g. if you add the following line to your source code, then it will display as .Net Framework 4.8:

```c#
[assembly: System.Runtime.Versioning.TargetFramework(".NETFramework,Version=v4.8", FrameworkDisplayName = ".NET Framework 4.8")]
```

But I don't think this attribute tells you which framework you're actually targeting when you're building the code yourself using `CSharpCompilation`.

**Reply by huoyaoyuan** [reply]
ILSpy seems displays target framework correctly.

**Reply by jasonmalinowski** [reply]
> But I don't think this attribute tells you which framework you're actually targeting when you're building the code yourself using CSharpCompilation.

Indeed; that attribute is produced by MSBuild and handed to the compiler during a regular project build. If you're building your own Compilations and emitting them, that would be up to you to replicate that logic.

**Comment by leet101** [comment, marked as answer]
https://github.com/dotnet/roslyn/discussions/57028#discussioncomment-1500743
